src_lockfree/htbl/htbl.body.ml

```diff
@@ -0,0 +1,617 @@
+  | Nil -> Nil_with_size { size_modifier = Size.used_once }
+  | Cons r ->
+      if t r.key land mask = chk then
+        Cons_with_size
+          {
+            key = r.key;
+            value = r.value;
+            size_modifier = Size.used_once;
+            rest = filter_ t mask chk r.rest;
+          }
+      else filter_fst t mask chk r.rest
+
+let filter t mask chk :
+    ('a, 'b, [ `Cons_with_size | `Nil_with_size ]) tdt ->
+    ('a, 'b, [> `Cons_with_size | `Nil_with_size ]) tdt = function
+  | Nil_with_size s -> Nil_with_size s
+  | Cons_with_size r -> begin
+      if t r.key land mask = chk then
+        Cons_with_size { r with rest = filter_ t mask chk r.rest }
+      else filter_fst t mask chk r.rest
+    end
+
+let rec split_all r target i t step =
+  let i = (i + step) land (Atomic_array.length r.buckets - 1) in
+  let spine = take_at Backoff.default r.size r.buckets i in
+  let high = Atomic_array.length r.buckets in
+  let after_lo = filter r.hash high 0 spine in
+  let after_hi = filter r.hash high high spine in
+  let (B before_lo) = Atomic_array.unsafe_fenceless_get target i in
+  let (B before_hi) = Atomic_array.unsafe_fenceless_get target (i + high) in
+  (* The [before_lo] and [before_hi] values are physically different for each
+     resize and so checking that the resize has not finished is sufficient to
+     ensure that the [compare_and_set] below does not disrupt the next
+     resize. *)
+  Atomic.get t == r
+  && begin
+       begin
+         match before_lo with
+         | Resize _ ->
+             Atomic_array.unsafe_compare_and_set target i (B before_lo)
+               (B after_lo)
+             |> ignore
+         | Nil_with_size _ | Cons_with_size _ -> ()
+       end;
+       begin
+         match before_hi with
+         | Resize _ ->
+             Atomic_array.unsafe_compare_and_set target (i + high) (B before_hi)
+               (B after_hi)
+             |> ignore
+         | Nil_with_size _ | Cons_with_size _ -> ()
+       end;
+       i = 0 || split_all r target i t step
+     end
+
+(* *)
+
+let[@tail_mod_cons] rec merge rest = function
+  | Nil -> rest
+  | Cons r -> Cons { r with rest = merge rest r.rest }
+
+let merge size (rest : ('a, 'b, [ `Nil_with_size | `Cons_with_size ]) tdt) :
+    ('a, 'b, [ `Nil_with_size | `Cons_with_size ]) tdt ->
+    ('a, 'b, [ `Nil_with_size | `Cons_with_size ]) tdt = function
+  | Nil_with_size r ->
+      (* Each size_modifier that is going to be removed need to be applied *)
+      if r.size_modifier != Size.used_once then
+        Size.update_once size r.size_modifier;
+      rest
+  | Cons_with_size r -> begin
+      match rest with
+      | Nil_with_size r' ->
+          if r'.size_modifier != Size.used_once then
+            Size.update_once size r'.size_modifier;
+          Cons_with_size r
+      | Cons_with_size r' -> begin
+          if r'.size_modifier != Size.used_once then
+            Size.update_once size r'.size_modifier;
```

Comment:
We can actually perform an optimization here to reduce code size:
```suggestion
      begin
        match rest with
        | Nil_with_size r' ->
            if r'.size_modifier != Size.used_once then
              Size.update_once size r'.size_modifier;
        | Cons_with_size r' -> begin
            if r'.size_modifier != Size.used_once then
              Size.update_once size r'.size_modifier;
      end;
      match rest with
      | Nil_with_size r' ->
          Cons_with_size r
      | Cons_with_size r' -> begin
```
This may seem like it would generate more code, but the compiler can actually optimize the code inside the `begin ... end` such that the size update is done for either case with just a single copy of the code, because the `size_modifier` field is at the same offset in both cases and so no pattern matching is required.